tests/integration/high_availability/test_self_healing.py

```diff
@@ -0,0 +1,172 @@
+
+    # ensure all units in the cluster are online
+    assert await ensure_n_online_mysql_members(
+        ops_test, 3
+    ), "The deployed mysql application is not fully online"
+
+    mysql_pid = await get_process_pid(ops_test, primary.name, "mysql", "mysqld")
+    assert mysql_pid > 0, "mysql process id is not positive"
+
+    await send_signal_to_pod_container_process(
+        ops_test,
+        primary.name,
+        "mysql",
+        "mysqld",
+        "SIGSTOP",
+    )
+
+    mysql_process_stat_after_sigstop = await get_process_stat(
+        ops_test, primary.name, "mysql", "mysqld"
+    )
+    assert "T" in mysql_process_stat_after_sigstop, "mysql process is not stopped after sigstop"
+
+    remaining_online_units = [
+        unit
+        for unit in ops_test.model.applications[mysql_application_name].units
+        if unit.name != primary.name
+    ]
+
+    for attempt in Retrying(stop=stop_after_delay(60), wait=wait_fixed(10)):
+        with attempt:
+            assert await ensure_n_online_mysql_members(
+                ops_test, 2, remaining_online_units
+            ), "The deployed mysql application does not have two online nodes"
+
+            new_primary = await get_primary_unit(
+                ops_test, remaining_online_units[0], mysql_application_name
+            )
+            assert primary.name != new_primary.name, "new mysql primary was not elected"
+
+    # since endpoints (in database relation databag) cannot get updated
+    # (as sigstops are not detected in charm code), insert some random data in the
+    # remaining units and ensure you can query them as well
+    database_name, table_name = "test-freeze-db-process", "data"
+    await insert_data_into_mysql_and_validate_replication(
+        ops_test, database_name, table_name, remaining_online_units
+    )
+    await clean_up_database_and_table(ops_test, database_name, table_name)
+
+    await send_signal_to_pod_container_process(
+        ops_test,
+        primary.name,
+        "mysql",
+        "mysqld",
+        "SIGCONT",
+    )
+
+    mysql_process_stat_after_sigcont = await get_process_stat(
+        ops_test, primary.name, "mysql", "mysqld"
+    )
+    assert (
+        "T" not in mysql_process_stat_after_sigcont
+    ), "mysql process is not started after sigcont"
+    assert (
+        "R" in mysql_process_stat_after_sigcont or "S" in mysql_process_stat_after_sigcont
+    ), "mysql process not running or sleeping after sigcont"
+
+    new_mysql_pid = await get_process_pid(ops_test, primary.name, "mysql", "mysqld")
+    assert (
+        new_mysql_pid == mysql_pid
+    ), "mysql process id is not the same as it was before process was stopped"
+
+    primary_after_sigstop = await get_primary_unit(
+        ops_test, remaining_online_units[0], mysql_application_name
+    )
+    assert new_primary.name == primary_after_sigstop.name, "mysql primary changed after sigstop"
+
+    assert await ensure_n_online_mysql_members(
```

Comment:
you'll want to be sure that you verify all units are in the same replica set, does this check do this?

Reply:
yep! since this helper method uses `get-cluster-status`, it will by default ensure that the N online members are part of the same cluster